tsdb/engine/tsm1/compact.go

```diff
@@ -573,7 +573,16 @@ func (c *Compactor) write(path string, iter KeyIterator) (err error) {
 		}
 
 		// Write the key and value
-		if err := w.WriteBlock(key, minTime, maxTime, block); err != nil {
+		err = w.WriteBlock(key, minTime, maxTime, block)
+		if err == ErrMaxBlocksExceeded {
+			if err := w.WriteIndex(); err != nil {
+				return err
+			}
+
+			return ErrMaxBlocksExceeded
```

Comment:
Also, stylistically I prefer grouping the entire thing together so the `err` is confined to a specific block:

```go
if err := w.WriteBlock(key, minTime, maxTime, block); err == ErrMaxBlocksExceeded {
	if err := w.WriteIndex(); err != nil {
		return err
	}
	return ErrMaxBlocksExceeded
} else if err != nil {
	return err
}
```